File: csharp/Knuckeotide.cs
using System.Text;

public class Knuckeotide : Benchmark
{
    private string _seq = "";
    private StringBuilder _resultBuilder;

    public override uint Checksum => Helper.Checksum(_resultBuilder.ToString());
    public override string TypeName => "CLBG::Knuckeotide";

    public Knuckeotide()
    {
        _resultBuilder = new StringBuilder();
    }

    public override void Prepare()
    {
        var fasta = new Fasta();
        fasta._n = ConfigVal("n");
        fasta.Prepare();
        fasta.Run(0);
        string result = fasta.GetResult();

        var seqBuilder = new StringBuilder();
        bool three = false;

        using (var reader = new StringReader(result))
        {
            string? line;
            while ((line = reader.ReadLine()) != null)
            {
                if (line.StartsWith(">THREE"))
                {
                    three = true;
                    continue;
                }
                if (three) seqBuilder.Append(line.Trim());
            }
        }

        _seq = seqBuilder.ToString();
    }

    private (int n, Dictionary<string, int> table) Frequency(string seq, int length)
    {
        int n = seq.Length - length + 1;
        var table = new Dictionary<string, int>();

        for (int i = 0; i < n; i++)
        {
            string sub = seq.Substring(i, length);
            table[sub] = table.GetValueOrDefault(sub, 0) + 1;
        }

        return (n, table);
    }

    private void SortByFreq(string seq, int length)
    {
        var (n, table) = Frequency(seq, length);

        var sorted = table.OrderByDescending(kv => kv.Value);

        foreach (var kv in sorted)
        {
            double freq = (kv.Value * 100.0) / n;
            _resultBuilder.AppendFormat("{0} {1:F3}\n", kv.Key.ToUpperInvariant(), freq);
        }
        _resultBuilder.Append('\n');
    }

    private void FindSeq(string seq, string s)
    {
        var (n, table) = Frequency(seq, s.Length);
        int count = table.GetValueOrDefault(s, 0);
        _resultBuilder.AppendFormat("{0}\t{1}\n", count, s.ToUpperInvariant());
    }

    public override void Run(long IterationId)
    {
        for (int i = 1; i <= 2; i++) SortByFreq(_seq, i);

        string[] patterns = { "ggt", "ggta", "ggtatt", "ggtattttaatt", "ggtattttaatttatagt" };
        foreach (var pattern in patterns) FindSeq(_seq, pattern);
    }
}

File: fsharp/Knuckeotide.fs
namespace Benchmarks

open System
open System.Collections.Generic
open System.IO
open System.Text

type Knuckeotide() =
    inherit Benchmark()

    let mutable n = 0L
    let mutable seq = ""
    let resultBuilder = StringBuilder()
    let mutable checksum = 0u

    let frequency (seq: string) length =
        let n = seq.Length - length + 1
        let table = Dictionary<string, int>()

        for i in 0 .. n - 1 do
            let sub = seq.Substring(i, length)
            match table.TryGetValue(sub) with
            | true, count -> table.[sub] <- count + 1
            | false, _ -> table.[sub] <- 1

        n, table

    let sortByFreq seq length =
        let n, table = frequency seq length

        let sorted = table |> Seq.sortByDescending (fun kv -> kv.Value)

        for kv in sorted do
            let freq = (float kv.Value * 100.0) / float n
            resultBuilder.AppendFormat("{0} {1:F3}\n", kv.Key.ToUpperInvariant(), freq) |> ignore

        resultBuilder.Append('\n') |> ignore

    let findSeq seq (s: string) =
        let n, table = frequency seq s.Length
        let count = match table.TryGetValue(s) with | true, c -> c | false, _ -> 0
        resultBuilder.AppendFormat("{0}\t{1}\n", count, s.ToUpperInvariant()) |> ignore

    override _.Checksum = 
        Helper.Checksum(resultBuilder.ToString())
    override this.Name = "CLBG::Knuckeotide"

    override _.Prepare() =
        n <- Helper.Config_i64("CLBG::Knuckeotide", "n")
        resultBuilder.Clear() |> ignore
        checksum <- 0u

        let fastaResult = FastaShared.generateFastaSequence n

        let seqBuilder = StringBuilder()
        let mutable three = false

        use reader = new StringReader(fastaResult)
        let rec readLines() =
            match reader.ReadLine() with
            | null -> ()
            | line ->
                if line.StartsWith(">THREE") then
                    three <- true
                elif three && not (line.StartsWith('>')) then
                    seqBuilder.Append(line.Trim()) |> ignore
                readLines()

        readLines()
        seq <- seqBuilder.ToString()

    override _.Run(IterationId: int64) =

        for i in 1 .. 2 do
            sortByFreq seq i

        let patterns = [|
            "ggt"; "ggta"; "ggtatt"; "ggtattttaatt"; "ggtattttaatttatagt"
        |]

        for pattern in patterns do
            findSeq seq pattern